### PR TITLE
Feature/offline email move

### LINKED ISSUE #382: Offline moving of mails
Ich als Kunde möchte E-Mails offline bearbeiten (löschen, in Ordner verschieben) können, damit ich auch offline produktiv sein kann. Dabei sollen die Änderungen wieder auf den Server synchronisiert werden.​

**Akzeptanzkriterium**:​ E-Mails müssen offline kategorisiert werden können. Dies muss am Ende alles an den Server zurückgemeldet werden, um die Änderungen damit zu synchronisieren.​

_Schätzung: L_

### PR DESCRIPTION
**Pull request by blackjack2693**
## What does this change?

The moving of emails to another folder is now working offline

Co-authored-by: *Johann Schulze Tast*
Co-authored-by: *christian Helms*

CLOSES #382
## Decisions / Choices I made

the implementation of a new class called ICMovedMailStorage was necessary to make sure there are no interferences with the server version while updating. the updating is now only possible if you are online, otherwise there could be inconsistencies. this is also the reason update is no longer part of the queue in CommandStorage. 

the methods to determine the highest and lowest UID in ICFolder had to be changed not to care about the moved emails, the server does not know about when being offline. otherwise the moved emails are no longer classified as moved but as normal emails of the folder they are in. 

## Checklist

- [x] The branch implementing the login changes should be merged before this one, this one based on it 
- [x] TEST IT locally to ensure there are no bugs in this version

## Note
Pull request will have to be merged manually. To prevent the disaster from last time:
- first merge `develop` into this branch (do **not** amend merge commits and be sure to find **all** merge conflicts before committing)
- then use GitHub to do the usual squash-and-merge

### CODE REVIEW

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st [hunk 1]
```diff
@@ -6,7 +6,10 @@ checkForBadRequest
   
 	(ICParser responseIsError: self stream lastResponse) ifTrue: [
 		self fatalErrorOccured: true.
-		self inform: 'Wrong username or password'.
+		BreakPoint break.
+		Transcript show: 'error username'; cr.
+		self inform: 'Wrong username or passwordddddddddd'.
+		Transcript show: self stream lastResponse; cr.
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st]: remove debugging stuff

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/isCommandStorageEmpty.st [hunk 2]
```diff
@@ -0,0 +1,4 @@
+move
+isCommandStorageEmpty
+
+	^ self storedCommands isEmptyFor: self accountInfo hash
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/isCommandStorageEmpty.st]: we really need to find a solution for the hash method (either make it work again or use another way of identifying accounts)

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st [hunk 3]
```diff
@@ -1,14 +1,4 @@
...
-	conOK := true.
-	
-	self storedCommands add: aCommand for: self accountInfo hash.
-	
-	[conOK and: (self storedCommands isEmptyFor: hash) not] whileTrue: [
-			[self sendDirectCommand: (self storedCommands getNextCommandFor: hash).
-			self storedCommands popNextCommandFor: hash]
-		on: Error do: [conOK := false]]
\ No newline at end of file
+sendCommand: aCommand
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st:2]: add newline :sweat_smile:

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st [hunk 4]
```diff
@@ -0,0 +1,2 @@
+connection
+sendNextCommands    | hash conOK |    hash := self accountInfo hash.  conOK := true.    [conOK and: (self storedCommands isEmptyFor: hash) not] whileTrue: [    [self sendDirectCommand: (self storedCommands getNextCommandFor: hash).    self storedCommands popNextCommandFor: hash]   on: Error do: [conOK := false]]
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st]: format method so that it isn't all in one line

Also, I noticed that ICCommandStorage has an unsqueakish method name (`getNextCommand`) in the current version on `develop` (see other comment). Maybe we could also change that somewhere.

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/moveEmail.to..st [hunk 5]
```diff
@@ -1,10 +1,20 @@
 protocol handling - move
-moveEmail: anEmail to: aFolder
-
-	self connectionHandler select: anEmail folder path.
-	self deleteMailLocally: anEmail.
+moveEmail: anEmail to: aFolderName
 	
-	self resetDeletedFlagWhenRecovering: anEmail.
+	"has to be refactored immediately"
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICEndPoint.class/instance/moveEmail.to..st]: maybe delete that line ^^

File: packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st [hunk 6]
```diff
@@ -1,4 +1,4 @@
 action - remove emails
 removeMail: anICEmail
 
-	self emails remove: anICEmail.
\ No newline at end of file
+	self emails remove: anICEmail ifAbsent: [ Transcript show: 'ICFolder removeMail is not there' ]
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st]: remove transcript output

File: packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st [hunk 7]
```diff
@@ -1,10 +1,16 @@
 action - update emails
 startUpdateProcess: anUID
-
-	(self folderName = 'root') ifTrue: [^true].
 	
+	| movedEmails acc |
+	acc := self accountName.
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st]: where do you use acc in this method? :sweat_smile:

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st [hunk 8]
```diff
@@ -0,0 +1,10 @@
+as yet unclassified
+deleteEmailsFromFolder: aFolder
+	"comment stating purpose of message"
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st]: remove line

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st [hunk 9]
```diff
@@ -0,0 +1,6 @@
+as yet unclassified
+getSetForAccount: anAccountName andFolder: aFolder
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st]: `get` prefix in method name is "unsqueakish" (apparently)

also shares lots of code with `addMail:forAccount:` (should probably be used in that method)

**Review by laugengebaeck — COMMENTED** [review]
Looks good and I'm happy that this also finally works now :partying_face: However, there are some minor issues that need to be addressed.

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/isCommandStorageEmpty.st (hunk 2 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/isCommandStorageEmpty.st]: CH suggested using the names of the accounts, but we need to enforce it. This is currently done in the MovedMailStorage

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/moveEmail.to..st (hunk 5 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICEndPoint.class/instance/moveEmail.to..st]: done

File: packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st (hunk 6 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st]: done

File: packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st (hunk 7 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st]: nowhere :D will be gone

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st (hunk 8 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st]: done

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st (hunk 9 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st]: anyone any suggestions for a better name? :D

Comment (laugengebaeck) [reply packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st]: `movedMailsForAccount:andFolder:` maybe?

**Comment by laugengebaeck** [comment]
Also, please write a short class comment for the `ICMovedEmailStorage` :)

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st (hunk 1 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st]: done

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st (hunk 4 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st]: done

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st (hunk 9 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/getSetForAccount.andFolder..st]: good idea, done

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/throwErrorIfNotLoggedIn.st [hunk 10]
```diff
@@ -1,8 +1,9 @@
 login
 throwErrorIfNotLoggedIn
-	
+
 	(self checkConnected) ifFalse: [
 		self fatalErrorOccured: true.
-		self inform: 'Cannot connect to the server'. 
+		self inform: 'Cannot connect to the server. Please check wether you are connected to the internet'.
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/throwErrorIfNotLoggedIn.st]:
```suggestion
		self inform: 'Cannot connect to the server. Please check whether you are connected to the internet'. 
```

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/README.md [hunk 11]
```diff
@@ -0,0 +1,7 @@
+An ICMovedEmailStorage is a dictionary of dictionaries containing sets of emails.
+it is also a *singleton*.
+The first dictionary represents the account / endpoint, the second one the folder the mail belongs to. 
+The mails were typically moved to the given folder and saved here while the clinet is offline.
```
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICMovedEmailStorage.class/README.md]:
```suggestion
The mails were typically moved to the given folder and saved here while the client is offline. 
```

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/updateProcess.st [hunk 12]
```diff
@@ -2,11 +2,10 @@ action - emails
...
-	[self emailAccounts do: [:anICEndpoint | successful := ( successful and: (
-    anICEndpoint rootFolder updateAndShowProgressWith: [self progressCallback]))]]
-  on: ConnectionClosed
+	successful := false.
+	[self emailAccounts do: [:anICEndpoint | successful := ( successful and: (anICEndpoint rootFolder updateAndShowProgressWith: [self progressCallback]))]]
+  on: NetworkError
 	do: [:e | self inform: e asString].
   
-	successful ifTrue: [self updateSynchTime].
+	successful ifFalse: [self updateSynchTime].
```
Comment (laugengebaeck) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/updateProcess.st]: Why was this changed? Also, that variable is now named misleadingly.

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkConnected.st [hunk 13]
```diff
@@ -1,8 +1,8 @@
 connection
 checkConnected
 	
-	(self fatalErrorOccured)
-		ifTrue: [^ false].
+	"(self fatalErrorOccured)
+		ifTrue: [^ false]."
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkConnected.st]: Please remove comments

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/select.direct..st [hunk 14]
```diff
@@ -0,0 +1,6 @@
+fetch
+select: aFolderName direct: aBool
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/select.direct..st]: Maybe call this method `select: aFolderName directly: aBool` instead. I find it a bit confusing at the moment that it gets a boolean as a parameter.

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st [hunk 15]
```diff
@@ -0,0 +1,10 @@
+connection
+sendNextCommands   
+	
+	| hash conOK |
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendNextCommands.st]: Please rename `conOK` to `isConnected` or something similar

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/testAccountWith..st [hunk 16]
```diff
@@ -2,6 +2,5 @@ login
 testAccountWith: anAccountInfo
 
 	"self connectAndLoginWith: anAccountInfo.
- 
-	^ self fatalErrorOccured not"
-	^ true
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICEndPoint.class/instance/testAccountWith..st]: Please remove comments

File: packages/IMAPClient-Core.package/ICFolder.class/instance/highestUniqueID.st [hunk 17]
```diff
@@ -1,6 +1,7 @@
 accessing
 highestUniqueID
 	
+	| movedEmailSet |
 	(self emails) ifEmpty: [^ nil].
-		
-	^ (self emails detectMax: [:mail | mail uniqueID asNumber]) uniqueID
\ No newline at end of file
+	movedEmailSet := ICMovedEmailStorage new movedMailsForAccount: self accountName andFolder: self.
+	^ (self emails detectMax: [:mail | (movedEmailSet includes: mail) ifFalse: [mail uniqueID asNumber]ifTrue: [-1]]) uniqueID
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICFolder.class/instance/highestUniqueID.st:7]: -1 is a magic number

File: packages/IMAPClient-Core.package/ICFolder.class/instance/lowestUniqueID.st [hunk 18]
```diff
@@ -1,6 +1,7 @@
 accessing
 lowestUniqueID
 	
+	| movedEmailSet |
 	(self emails) ifEmpty: [^ nil].
-		
-	^ (self emails detectMin: [:mail | mail uniqueID asNumber]) uniqueID
\ No newline at end of file
+	movedEmailSet := ICMovedEmailStorage new movedMailsForAccount: self accountName andFolder: self.
+	^ (self emails detectMin: [:mail | (movedEmailSet includes: mail) ifFalse: [mail uniqueID asNumber] ifTrue: [9999999]] ) uniqueID
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICFolder.class/instance/lowestUniqueID.st:7]: 999999 is a magic number

File: packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st [hunk 19]
```diff
@@ -1,4 +1,4 @@
 action - remove emails
-removeMail: anICEmail
+removeMail: anEmail
 
-	self emails remove: anICEmail.
\ No newline at end of file
+	self emails remove: anEmail ifAbsent: []
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st]: what does ifAbsent: [] do?

File: packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st [hunk 20]
```diff
@@ -1,10 +1,15 @@
 action - update emails
 startUpdateProcess: anUID
-
-	(self folderName = 'root') ifTrue: [^true].
 	
+	| movedEmails |
+	movedEmails := ICMovedEmailStorage new movedMailsForAccount: self accountName andFolder: self.
+	(self folderName = 'root') ifTrue: [ ^true].
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICFolder.class/instance/startUpdateProcess..st]: `[^ true]`

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/movedMailsForAccount.andFolder..st [hunk 21]
```diff
@@ -0,0 +1,6 @@
+accessing
+movedMailsForAccount: anAccountName andFolder: aFolder
+	
+	 | accountDictionary |
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/movedMailsForAccount.andFolder..st]: new line

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/movedMailsForAccount.andFolder..st [hunk 22]
```diff
@@ -0,0 +1,6 @@
...
+	accountDictionary := self at: anAccountName ifAbsent: [self at: anAccountName put: Dictionary new].
+	^(accountDictionary at:  aFolder ifAbsent: [ accountDictionary at: aFolder put: Set new.])
```
Comment (sm1lla) [inline packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/movedMailsForAccount.andFolder..st]: Space after `^`

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/switchToNewEndpoint.st [hunk 23]
```diff
@@ -3,8 +3,8 @@ switchToNewEndpoint
 
 	
 	self selectedEndpoint ifNil: [^self].	
-	self folderWith: self selectedEndpoint.
-	
+	"self folderWith: self selectedEndpoint."
```
Comment (sm1lla) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/switchToNewEndpoint.st]: Please remove comment

**Review by sm1lla — APPROVED** [review]
Works nicely! Just some comments on the style.

**Review by sm1lla — CHANGES_REQUESTED** [review]
But please address the failing tests.

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st [hunk 24]
```diff
@@ -6,7 +6,7 @@ checkForBadRequest
   
 	(ICParser responseIsError: self stream lastResponse) ifTrue: [
 		self fatalErrorOccured: true.
-		self inform: 'Wrong username or password'.
+		self inform: 'responseError in checkForBadRequest'.
```
Comment (elenagensch) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st:9]: We should keep in mind that it might be confusing for the users when they login with wrong username or password and get this message (instead of wrong username or password). Error messages for users should be thrown in higher layers (when you know what operation failed)

Clearer error messages could be a good first issue for the next group.

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/select.direct..st (hunk 14 above)
Comment (elenagensch) [reply packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/select.direct..st]: I agree

File: packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st [hunk 25]
```diff
@@ -0,0 +1,9 @@
+mailHandling
+deleteEmailsFromFolder: aFolder
+
+	| accountName emailSet |
+	
+	accountName := aFolder accountName.
+	emailSet := self movedMailsForAccount: accountName andFolder: aFolder.
+	emailSet do: [:anEmail | aFolder endpoint deleteMailLocally: anEmail].
+
```
Comment (elenagensch) [inline packages/IMAPClient-Core.package/ICMovedEmailStorage.class/instance/deleteEmailsFromFolder..st]: newline

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st [hunk 26]
```diff
@@ -0,0 +1,10 @@
+action - emails
+login: anEndpoint
+
+	| continueAsking |
+	continueAsking := true. 
+	[continueAsking] whileTrue: [
+		anEndpoint accountInfo password: (Project uiManager requestPassword: 'Please enter your password!').
+		anEndpoint accountInfo password ifEmpty: [^false].
+		continueAsking := (self passwordPromtCallback: anEndpoint) not].
+	^ true.
```
Comment (elenagensch) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st]: remove point at the end

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/passwordPromtCallback..st [hunk 27]
```diff
@@ -1,13 +1,9 @@
...
-		(account accountName) = (anAccountInfo accountName)].
-	(self uninitializedAccounts at: nextAccount) ifNotNil: [self uninitializedAccounts removeAt: nextAccount].
+	self folderWith: anEndpoint.
+	anEndpoint preventTimeout.
 	
-	(self uninitializedAccounts isEmpty)
-		ifTrue: [self accountsComplete]
-		ifFalse: [ICPasswordDialog openWith: (self uninitializedAccounts first) and: self]
\ No newline at end of file
+	^true
```
Comment (elenagensch) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/passwordPromtCallback..st]: ^ true

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st [hunk 28]
```diff
@@ -1,14 +1,5 @@
...
-sendCommand: aCommand
-	
-	| hash conOK |
-	
-	hash := self accountInfo hash.
-	conOK := true.
-	
-	self storedCommands add: aCommand for: self accountInfo hash.
-	
-	[conOK and: (self storedCommands isEmptyFor: hash) not] whileTrue: [
```
Comment (christian-helms) [inline packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st:10]: We should start identifying accounts by their name or a dedicated identifier instead of some weird hash.

File: packages/IMAPClient-Core.package/ICFolder.class/instance/lowestUniqueID.st (hunk 18 above)
Comment (christian-helms) [reply packages/IMAPClient-Core.package/ICFolder.class/instance/lowestUniqueID.st:7]: Yeah, that's weird.

File: packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st (hunk 19 above)
Comment (christian-helms) [reply packages/IMAPClient-Core.package/ICFolder.class/instance/removeMail..st]: Indeed it is unnecessary as OrderedCollection inherits from Collection which implements an element-based remove with a default specification for the case that the element is not found.

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/updateProcess.st (hunk 12 above)
Comment (christian-helms) [reply packages/IMAPClient-UI.package/ICFolderDialog.class/instance/updateProcess.st]: You're right. That's an obvious mistake.

**Review by christian-helms — APPROVED** [review]

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/updateProcess.st (hunk 12 above)
Comment (laugengebaeck) [reply packages/IMAPClient-UI.package/ICFolderDialog.class/instance/updateProcess.st]: Fixed/reverted :)

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st (hunk 24 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/checkForBadRequest.st:9]: good idea

File: packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st (hunk 28 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICConnectionHandler.class/instance/sendCommand..st:10]: i am now using the email address, this should be unique 😉

File: packages/IMAPClient-Core.package/ICFolder.class/instance/lowestUniqueID.st (hunk 18 above)
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICFolder.class/instance/lowestUniqueID.st:7]: it makes sure that while fetching new mails the moved ones, which could have the same UID as a mail already in the system, the added email is ignored.

**Review by sm1lla — APPROVED** [review]
